COMP3301 Operating Systems Architecture/a1/a1/ps.1
.\"     $OpenBSD: ps.1,v 1.139 2024/10/15 13:49:49 claudio Exp $
.\"     $NetBSD: ps.1,v 1.16 1996/03/21 01:36:28 jtc Exp $
.\"
.\" Copyright (c) 1980, 1990, 1991, 1993, 1994
.\"     The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)ps.1        8.3 (Berkeley) 4/18/94
.\"
.Dd $Mdocdate: October 15 2024 $
.Dt PS 1
.Os
.Sh NAME
.Nm ps
.Nd display process status
.Sh SYNOPSIS
.Nm ps
.Sm off
.Op Oo Fl Oc Cm AacefHhjkLlmrSTuvwxZ
.Sm on
.Op Fl M Ar core
.Op Fl N Ar system
.Op Fl O Ar fmt
.Op Fl o Ar fmt
.Op Fl p Ar pid
.Op Fl t Ar tty
.Op Fl U Ar user
.Op Fl W Ar swap
.Op Fl z Ar zone
.Sh DESCRIPTION
The
.Nm
utility displays information about active processes.
When given no options,
.Nm
prints information about processes of the current user that have a
controlling terminal.
.Pp
The information displayed is selected based on a set of keywords (and for
even more control, see the
.Fl L ,
.Fl O ,
and
.Fl o
options).
The default output format includes, for each process, the process's ID,
controlling terminal, state, CPU time (including both user and system time),
and associated command.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl A
Display information about processes for all users,
including those without controlling terminals.
.It Fl a
Display information about processes for all users with controlling terminals.
.It Fl c
Do not display full command with arguments, but only the
executable name.
This may be somewhat confusing; for example, all
.Xr sh 1
scripts will show as
.Dq sh .
.It Fl e
Display the environment as well.
.It Fl f
Arrange processes into a tree.
Each command is ordered and prefixed with indentation text
showing sibling and parent/child relationships.
If either of the
.Fl m
and
.Fl r
options are also used, they control how sibling processes are sorted relative
to each other.
.It Fl H
Also display information about kernel visible threads.
.It Fl h
Repeat the information header as often as necessary to guarantee one
header per page of information.
.It Fl j
Print information associated with the following keywords:
user, pid, ppid, pgid, sess, jobc, state, tt, time, and command.
.It Fl k
Also display information about kernel threads.
.It Fl L
List the set of available keywords.
This option should not be specified with other options.
.It Fl l
Display information associated with the following keywords:
uid, pid, ppid, cpu, pri, nice, vsz, rss, wchan, state, tt, time,
and command.
.It Fl M Ar core
Extract values associated with the name list from the specified core
instead of the running kernel.
.It Fl m
Sort by memory usage, instead of by start time ID.
.It Fl N Ar system
Extract the name list from the specified system instead of the running kernel.
.It Fl O Ar fmt
Add the information associated with the space or comma separated list
of keywords specified, after the process ID,
in the default information
display.
Keywords may be appended with an equals sign
.Pq Sq =
and a string.
This causes the printed header to use the specified string instead of
the standard header.
.It Fl o Ar fmt
Display information associated with the space or comma separated list
of keywords specified.
Keywords may be appended with an equals sign
.Pq Sq =
and a string.
This causes the printed header to use the specified string instead of
the standard header.
.It Fl p Ar pid
Display information associated with the specified process ID.
.It Fl r
Sort by current CPU usage, instead of by start time ID.
.It Fl S
Change the way the process time is calculated by summing all exited
children to their parent process.
.It Fl T
Display information about processes attached to the device associated
with the standard input.
.It Fl t Ar tty
Display information about processes attached to the specified terminal
device.
.It Fl U Ar user
Only display processes belonging to username or UID
.Ar user .
.It Fl u
Display information associated with the following keywords:
user, pid, %cpu, %mem, vsz, rss, tt, state, start, time, and command.
The
.Fl u
option implies the
.Fl r
option.
.It Fl v
Display information associated with the following keywords:
pid, state, time, sl, re, pagein, vsz, rss, lim, tsiz,
%cpu, %mem, and command.
The
.Fl v
option implies the
.Fl m
option.
.It Fl W Ar swap
When not using the running kernel,
extract swap information from the specified file.
.It Fl w
Use 132 columns to display information, instead of the default, which
is the window size.
If the
.Fl w
option is specified more than once,
.Nm
will use as many columns as necessary without regard for window size.
.It Fl x
Display information about processes without controlling terminals.
.It Fl Z
Prepends the displayed information with the zone field.
.It Fl z Ar zone
Display information about processes running in the specified zone.
.El
.Sh KEYWORDS
The following is a complete list of the available keywords
and their meanings.
Several of them have aliases,
which are also noted.
.Bl -tag -width "sigignoreXX" -offset 3n
.It Cm %cpu
Alias:
.Cm pcpu .
The CPU utilization of the process; this is a decaying average over up to
a minute of previous (real) time.
Since the time base over which this is computed varies (since processes may
be very young), it is possible for the sum of all
.Cm %cpu
fields to exceed 100%.
.It Cm %mem
Alias:
.Cm pmem .
The percentage of real memory used by this process.
.It Cm acflag
Alias:
.Cm acflg .
Accounting flag.
.It Cm command
Alias:
.Cm args .
Command and arguments.
.It Cm cpu
Short-term CPU usage factor (for scheduling).
.It Cm cpuid
CPU ID (zero on single processor systems).
.It Cm cwd
Current working directory.
.It Cm dsiz
Data size, in Kilobytes.
.It Cm etime
Elapsed time since the process was started.
.It Cm flags
Alias:
.Cm f .
The thread flags (in hexadecimal), as defined in the include file
.In sys/proc.h :
.Bd -literal
P_INKTR           0x1 writing ktrace(2) record
P_PROFPEND        0x2 this thread needs SIGPROF
P_ALRMPEND        0x4 this thread needs SIGVTALRM
P_SIGSUSPEND      0x8 need to restore before-suspend mask
P_CANTSLEEP      0x10 this thread is not permitted to sleep
P_SINTR          0x80 sleep is interruptible
P_SYSTEM        0x200 system process: no sigs, stats, or
                      swapping
P_TIMEOUT       0x400 timing out during sleep
P_WEXIT        0x2000 working on exiting
P_OWEUPC       0x8000 profiling sample needs recording
P_SUSPSINGLE  0x80000 need to suspend for single threading
P_THREAD    0x4000000 not the original thread
P_SUSPSIG   0x8000000 stopped because of a signal
P_CPUPEG   0x40000000 do not move to another cpu
.Ed
.It Cm gid
Effective group.
.It Cm group
Text name of effective group ID.
.It Cm inblk
Alias:
.Cm inblock .
Total blocks read.
.It Cm jobc
Job control count.
.It Cm ktrace
Tracing flags.
.It Cm ktracep
Tracing vnode.
.It Cm lim
The soft limit on memory used, specified via a call to
.Xr setrlimit 2 .
.It Cm logname
Alias:
.Cm login .
Login name of user who started the process.
.It Cm lstart
The exact time the command started, using the
.Dq %c
format described in
.Xr strftime 3 .
.It Cm majflt
Total page faults.
.It Cm maxrss
Maximum resident set size (in 1024 byte units).
.It Cm minflt
Total page reclaims.
.It Cm msgrcv
Total messages received (reads from pipes/sockets).
.It Cm msgsnd
Total messages sent (writes on pipes/sockets).
.It Cm nice
Alias:
.Cm ni .
The process scheduling increment (see
.Xr setpriority 2 ) .
.It Cm nivcsw
Total involuntary context switches.
.It Cm nsigs
Alias:
.Cm nsignals .
Total signals taken.
.It Cm nswap
Total swaps in/out.
.It Cm nvcsw
Total voluntary context switches.
.It Cm nwchan
Wait channel (as an address).
.It Cm oublk
Alias:
.Cm oublock .
Total blocks written.
.It Cm p_ru
Resource usage (valid only for zombie processes).
.It Cm paddr
Swap address.
.It Cm pagein
Pageins (same as
.Cm majflt ) .
.It Cm pgid
Process group number.
.It Cm pid
Process ID.
.It Cm pledge
Comma separated list of active
.Xr pledge 2
promises.
.It Cm ppid
Parent process ID.
.It Cm pri
Scheduling priority.
.It Cm procflags
The process flags (in hexadecimal), as defined in the include file
.In sys/proc.h :
.Bd -literal
PS_CONTROLT            0x1 process has a controlling
                           terminal
PS_EXEC                0x2 process called exec(3)
PS_INEXEC              0x4 process is doing an exec right
                           now
PS_EXITING             0x8 process is exiting
PS_SUGID              0x10 process had set ID privileges
                           since last exec
PS_SUGIDEXEC          0x20 last exec(3) was set[ug]id
PS_PPWAIT             0x40 parent is waiting for process
                           to exec/exit
PS_ISPWAIT            0x80 process is parent of PPWAIT
                           child
PS_PROFIL            0x100 process has started profiling
PS_TRACED            0x200 process is being traced
PS_WAITED            0x400 debugging process has waited
                           for child
PS_COREDUMP          0x800 busy coredumping
PS_SINGLEEXIT       0x1000 other threads must die
PS_SINGLEUNWIND     0x2000 other threads must unwind
PS_NOZOMBIE         0x4000 pid 1 waits for me instead of
                           dad
PS_STOPPING         0x8000 just stopped, need to send
                           SIGCHLD
PS_SYSTEM          0x10000 No signals, stats or swapping
PS_EMBRYO          0x20000 New process, not yet fledged
PS_ZOMBIE          0x40000 Dead and ready to be waited for
PS_NOBROADCASTKILL 0x80000 Process excluded from kill -1
PS_PLEDGE         0x100000 process has called pledge(2)
PS_WXNEEDED     0x00200000 process allowed to violate W^X
PS_EXECPLEDGE   0x00400000 has exec pledges
PS_ORPHAN       0x00800000 process is on an orphan list
PS_CHROOT       0x01000000 process is chrooted
PS_NOBTCFI      0x02000000 no Branch Target CFI
PS_CONTINUED    0x20000000 process continued from stopped state
                           but has not been waited for yet
PS_STOPPED      0x40000000 process is in stopped state
.Ed
.It Cm re
Core residency time (in seconds; 127 = infinity).
.It Cm rgid
Real group ID.
.It Cm rgroup
Text name of real group ID.
.It Cm rlink
Reverse link on run queue, or 0.
.It Cm rss
The real memory (resident set) size of the process (in 1024 byte units).
.It Cm rsz
Alias:
.Cm rssize .
Resident set size + (text size / text use count).
.It Cm rtable
Routing table.
.It Cm ruid
Real user ID.
.It Cm ruser
User name (from
.Cm ruid ) .
.It Cm sess
Session ID (PID of session leader).
.It Cm sig
Alias:
.Cm pending .

COMP3301 Operating Systems Architecture/a1/a1/ps.c
/*      $OpenBSD: ps.c,v 1.81 2024/05/18 13:08:09 sobrado Exp $ */
/*      $NetBSD: ps.c,v 1.15 1995/05/18 20:33:25 mycroft Exp $  */

/*-
 * Copyright (c) 1990, 1993, 1994
 *      The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/param.h>  /* NODEV */
#include <sys/types.h>
#include <sys/signal.h>
#include <sys/sysctl.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <sys/proc.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <sys/zones.h>

#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <kvm.h>
#include <locale.h>
#include <nlist.h>
#include <paths.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <limits.h>

#include "ps.h"

extern char *__progname;

struct varent *vhead;

int     eval;                   /* exit value */
int     sumrusage;              /* -S */
int     termwidth;              /* width of screen (0 == infinity) */
int     totwidth;               /* calculated width of requested variables */

int     needcomm, needenv, neednlist, commandonly;

enum sort { DEFAULT, SORTMEM, SORTCPU } sortby = DEFAULT;

static char     *kludge_oldps_options(char *);
static zoneid_t  zone_lookup(const char *);
static int       pscomp(const void *, const void *);
static void      scanvars(void);
static void      forest_sort(struct pinfo *, int);
static void      usage(void);

char dfmt[] = "pid tt state time command";
char tfmt[] = "pid tid tt state time command";
char jfmt[] = "user pid ppid pgid sess jobc state tt time command";
char lfmt[] = "uid pid ppid cpu pri nice vsz rss wchan state tt time command";
char   o1[] = "pid";
char   o2[] = "tt state time command";
char ufmt[] = "user pid %cpu %mem vsz rss tt state start time command";
char vfmt[] = "pid state time sl re pagein vsz rss lim tsiz %cpu %mem command";

kvm_t *kd;
int kvm_sysctl_only;

int
main(int argc, char *argv[])
{
        struct kinfo_proc *kp;
        struct pinfo *pinfo;
        struct varent *vent;
        struct winsize ws;
        dev_t ttydev;
        pid_t pid;
        uid_t uid;
        int all, ch, flag, i, fmt, lineno, nentries;
        int prtheader, showthreads, wflag, kflag, what, Uflag, xflg;
        int forest;
        char *nlistf, *memf, *swapf, *cols, errbuf[_POSIX2_LINE_MAX];
        int Zflag = 0;
        const char *zone = NULL;
        zoneid_t zoneid = -1;

        setlocale(LC_CTYPE, "");

        termwidth = 0;
        if ((cols = getenv("COLUMNS")) != NULL)
                termwidth = strtonum(cols, 1, INT_MAX, NULL);
        if (termwidth == 0 &&
            (ioctl(STDOUT_FILENO, TIOCGWINSZ, &ws) == 0 ||
            ioctl(STDERR_FILENO, TIOCGWINSZ, &ws) == 0 ||
            ioctl(STDIN_FILENO, TIOCGWINSZ, &ws) == 0) &&
            ws.ws_col > 0)
                termwidth = ws.ws_col - 1;
        if (termwidth == 0)
                termwidth = 79;

        if (argc > 1)
                argv[1] = kludge_oldps_options(argv[1]);

        all = fmt = prtheader = showthreads = wflag = kflag = Uflag = xflg = 0;
        pid = -1;
        uid = 0;
        forest = 0;
        ttydev = NODEV;
        memf = nlistf = swapf = NULL;
        while ((ch = getopt(argc, argv,
            "AaCcefgHhjkLlM:mN:O:o:p:rSTt:U:uvW:wxZz:")) != -1)
                switch (ch) {
                case 'A':
                        all = 1;
                        xflg = 1;
                        break;
                case 'a':
                        all = 1;
                        break;
                case 'C':
                        break;                  /* no-op */
                case 'c':
                        commandonly = 1;
                        break;
                case 'e':                       /* XXX set ufmt */
                        needenv = 1;
                        break;
                case 'f':
                        forest = 1;
                        break;
                case 'g':
                        break;                  /* no-op */
                case 'H':
                        showthreads = 1;
                        break;
                case 'h':
                        prtheader = ws.ws_row > 5 ? ws.ws_row : 22;
                        break;
                case 'j':
                        parsefmt(jfmt);
                        fmt = 1;
                        jfmt[0] = '\0';
                        break;
                case 'k':
                        kflag = 1;
                        break;
                case 'L':
                        showkey();
                        exit(0);
                case 'l':
                        parsefmt(lfmt);
                        fmt = 1;
                        lfmt[0] = '\0';
                        break;
                case 'M':
                        memf = optarg;
                        break;
                case 'm':
                        sortby = SORTMEM;
                        break;
                case 'N':
                        nlistf = optarg;
                        break;
                case 'O':
                        parsefmt(o1);
                        parsefmt(optarg);
                        parsefmt(o2);
                        o1[0] = o2[0] = '\0';
                        fmt = 1;
                        break;
                case 'o':
                        parsefmt(optarg);
                        fmt = 1;
                        break;
                case 'p':
                        pid = atol(optarg);
                        xflg = 1;
                        break;
                case 'r':
                        sortby = SORTCPU;
                        break;
                case 'S':
                        sumrusage = 1;
                        break;
                case 'T':
                        if ((optarg = ttyname(STDIN_FILENO)) == NULL)
                                errx(1, "stdin: not a terminal");
                        /* FALLTHROUGH */
                case 't': {
                        struct stat sb;
                        char *ttypath, pathbuf[PATH_MAX];

                        if (strcmp(optarg, "co") == 0)
                                ttypath = _PATH_CONSOLE;
                        else if (*optarg != '/') {
                                int r = snprintf(pathbuf, sizeof(pathbuf), "%s%s",
                                    _PATH_TTY, optarg);
                                if (r < 0 || r > sizeof(pathbuf))
                                        errx(1, "%s: too long\n", optarg);
                                ttypath = pathbuf;
                        } else
                                ttypath = optarg;
                        if (stat(ttypath, &sb) == -1)
                                err(1, "%s", ttypath);
                        if (!S_ISCHR(sb.st_mode))
                                errx(1, "%s: not a terminal", ttypath);
                        ttydev = sb.st_rdev;
                        break;
                }
                case 'U': {
                        int found = 0;

                        if (uid_from_user(optarg, &uid) == 0)
                                found = 1;
                        else {
                                const char *errstr;

                                uid = strtonum(optarg, 0, UID_MAX, &errstr);
                                if (errstr == NULL &&
                                    user_from_uid(uid, 1) != NULL)
                                        found = 1;
                        }
                        if (!found)
                                errx(1, "%s: unknown user", optarg);
                        Uflag = xflg = 1;
                        break;
                }
                case 'u':
                        parsefmt(ufmt);
                        sortby = SORTCPU;
                        fmt = 1;
                        ufmt[0] = '\0';
                        break;
                case 'v':
                        parsefmt(vfmt);
                        sortby = SORTMEM;
                        fmt = 1;
                        vfmt[0] = '\0';
                        break;
                case 'W':
                        swapf = optarg;
                        break;
                case 'w':
                        if (wflag)
                                termwidth = UNLIMITED;
                        else if (termwidth < 131)
                                termwidth = 131;
                        wflag = 1;
                        break;
                case 'x':
                        xflg = 1;
                        break;
                case 'Z':
                        Zflag = 1;
                        break;
                case 'z':
                        zone = optarg;
                        break;
                default:
                        usage();
                }
        argc -= optind;
        argv += optind;

#define BACKWARD_COMPATIBILITY
#ifdef  BACKWARD_COMPATIBILITY
        if (*argv) {
                nlistf = *argv;
                if (*++argv) {
                        memf = *argv;
                        if (*++argv)
                                swapf = *argv;
                }
        }

#endif

        if (nlistf == NULL && memf == NULL && swapf == NULL) {
                kd = kvm_openfiles(NULL, NULL, NULL, KVM_NO_FILES, errbuf);
                kvm_sysctl_only = 1;
        } else {
                kd = kvm_openfiles(nlistf, memf, swapf, O_RDONLY, errbuf);
        }
        if (kd == NULL)
                errx(1, "%s", errbuf);

        if (unveil(_PATH_DEVDB, "r") == -1 && errno != ENOENT)
                err(1, "unveil %s", _PATH_DEVDB);
        if (unveil(_PATH_DEV, "r") == -1 && errno != ENOENT)
                err(1, "unveil %s", _PATH_DEV);
        if (swapf)
                if (unveil(swapf, "r") == -1)
                        err(1, "unveil %s", swapf);
        if (nlistf)
                if (unveil(nlistf, "r") == -1)
                        err(1, "unveil %s", nlistf);
        if (memf)
                if (unveil(memf, "r") == -1)
                        err(1, "unveil %s", memf);
        if (pledge("stdio rpath getpw ps", NULL) == -1)
                err(1, "pledge");

        if (!fmt) {
                if (showthreads)
                        parsefmt(tfmt);
                else
                        parsefmt(dfmt);
        }
        if (Zflag)
                zonefmt();

        /* XXX - should be cleaner */
        if (!all && ttydev == NODEV && pid == -1 && !Uflag) {
                uid = getuid();
                Uflag = 1;
        }

        /*
         * scan requested variables, noting what structures are needed,
         * and adjusting header widths as appropriate.
         */
        scanvars();

        if (neednlist && !nlistread)
                (void) donlist();
        /*
         * get proc list
         */
        if (Uflag) {
                what = KERN_PROC_UID;
                flag = uid;
        } else if (ttydev != NODEV) {
                what = KERN_PROC_TTY;
                flag = ttydev;
        } else if (pid != -1) {
                what = KERN_PROC_PID;
                flag = pid;
        } else if (kflag) {
                what = KERN_PROC_KTHREAD;
                flag = 0;
        } else {
                what = KERN_PROC_ALL;
                flag = 0;
        }
        if (showthreads)
                what |= KERN_PROC_SHOW_THREADS;

        if (zone != NULL)
                zoneid = zone_lookup(zone);

        /*
         * select procs
         */
        kp = kvm_getprocs(kd, what, flag, sizeof(*kp), &nentries);
        if (kp == NULL)
                errx(1, "%s", kvm_geterr(kd));

        /*
         * print header
         */
        printheader();
        if (nentries == 0)
                exit(1);

        if ((pinfo = calloc(nentries, sizeof(struct pinfo))) == NULL)
                err(1, NULL);
        for (i = 0; i < nentries; i++)
                pinfo[i].ki = &kp[i];
        qsort(pinfo, nentries, sizeof(struct pinfo), pscomp);

        if (forest)
                forest_sort(pinfo, nentries);

        /*
         * for each proc, call each variable output function.
         */
        for (i = lineno = 0; i < nentries; i++) {
                if (zoneid != -1 && zoneid != pinfo[i].ki->p_zoneid)
                        continue;
                if (xflg == 0 && ((int)pinfo[i].ki->p_tdev == NODEV ||
                    (pinfo[i].ki->p_psflags & PS_CONTROLT ) == 0))
                        continue;
                if (showthreads && pinfo[i].ki->p_tid == -1)
                        continue;
                for (vent = vhead; vent; vent = vent->next) {
                        (vent->var->oproc)(&pinfo[i], vent);
                        if (vent->next != NULL)
                                (void)putchar(' ');
                }
                (void)putchar('\n');
                if (prtheader && lineno++ == prtheader - 4) {
                        (void)putchar('\n');
                        printheader();
                        lineno = 0;
                }
        }
        exit(eval);
}

static zoneid_t
zone_lookup(const char *zone)
{
        const char *errstr;
        zoneid_t z;

        z = zone_id(zone);
        if (z == -1) {
                if (errno != ESRCH)
                        err(1, "zone lookup");

                z = strtonum(zone, 0, 1024, &errstr);
                if (errstr != NULL)
                        errx(1, "zone lookup: %s", errstr);
        }

        return (z);
}

static void
scanvars(void)
{
        struct varent *vent;
        VAR *v;
        int i;

        for (vent = vhead; vent; vent = vent->next) {
                v = vent->var;
                i = strlen(v->header);
                if (v->width < i)
                        v->width = i;
                totwidth += v->width + 1;       /* +1 for space */
                if (v->flag & COMM)
                        needcomm = 1;
                if (v->flag & NLIST)
                        neednlist = 1;
        }
        totwidth--;
}

static int
pscomp(const void *v1, const void *v2)
{
        const struct pinfo *p1 = (const struct pinfo *)v1;
        const struct pinfo *p2 = (const struct pinfo *)v2;
        const struct kinfo_proc *kp1 = p1->ki;
        const struct kinfo_proc *kp2 = p2->ki;
        int i;
#define VSIZE(k) ((k)->p_vm_dsize + (k)->p_vm_ssize + (k)->p_vm_tsize)

        if (sortby == SORTCPU && (i = getpcpu(kp2) - getpcpu(kp1)) != 0)
                return (i);
        if (sortby == SORTMEM && (i = VSIZE(kp2) - VSIZE(kp1)) != 0)
                return (i);
        if ((i = kp1->p_tdev - kp2->p_tdev) == 0 &&
            (i = kp1->p_ustart_sec - kp2->p_ustart_sec) == 0)
                i = kp1->p_ustart_usec - kp2->p_ustart_usec;
        return (i);
}

/*
 * ICK (all for getopt), would rather hide the ugliness
 * here than taint the main code.
 *
 *  ps foo -> ps -foo
 *  ps 34 -> ps -p34
 *
 * The old convention that 't' with no trailing tty arg means the users
 * tty, is only supported if argv[1] doesn't begin with a '-'.  This same
 * feature is available with the option 'T', which takes no argument.
 */
static char *
kludge_oldps_options(char *s)
{
        size_t len;
        char *newopts, *ns, *cp;

        len = strlen(s);
        if ((newopts = ns = malloc(2 + len + 1)) == NULL)
                err(1, NULL);
        /*
         * options begin with '-'
         */
        if (*s != '-')
                *ns++ = '-';    /* add option flag */

        /*
         * gaze to end of argv[1]
         */
        cp = s + len - 1;
        /*
         * if last letter is a 't' flag with no argument (in the context
         * of the oldps options -- option string NOT starting with a '-' --
         * then convert to 'T' (meaning *this* terminal, i.e. ttyname(0)).
         */
        if (*cp == 't' && *s != '-')
                *cp = 'T';
        else {
                /*
                 * otherwise check for trailing number, which *may* be a
                 * pid.
                 */
                while (cp >= s && isdigit((unsigned char)*cp))
                        --cp;
        }
        cp++;
        memmove(ns, s, (size_t)(cp - s));       /* copy up to trailing number */
        ns += cp - s;
        /*
         * if there's a trailing number, and not a preceding 'p' (pid),
         * 't' (tty) or 'U' (user) flag,
         * then assume it's a pid and insert a 'p' flag.
         */
        if (isdigit((unsigned char)*cp) &&
            (cp == s || (cp[-1] != 't' && cp[-1] != 'p' && cp[-1] != 'U' &&
            (cp - 1 == s || cp[-2] != 't'))))
                *ns++ = 'p';
        /* and append the number */
        (void)strlcpy(ns, cp, newopts + len + 3 - ns);

        return (newopts);
}

static void
forest_sort(struct pinfo *ki, int items)
{
        int dst, lvl, maxlvl, n, ndst, nsrc, siblings, src;
        unsigned char *path;
        struct pinfo kn;

        /*
         * First, sort the entries by forest, tracking the forest
         * depth in the level field.
         */
        src = 0;
        maxlvl = 0;
        while (src < items) {
                if (ki[src].level) {
                        src++;
                        continue;
                }
                for (nsrc = 1; src + nsrc < items; nsrc++)
                        if (!ki[src + nsrc].level)
                                break;

                for (dst = 0; dst < items; dst++) {
                        if (ki[dst].ki->p_pid == ki[src].ki->p_pid)
                                continue;
                        if (ki[dst].ki->p_pid == ki[src].ki->p_ppid)
                                break;
                }

                if (dst == items) {
                        src += nsrc;
                        continue;
                }

                for (ndst = 1; dst + ndst < items; ndst++)
                        if (ki[dst + ndst].level <= ki[dst].level)
                                break;

                for (n = src; n < src + nsrc; n++) {
                        ki[n].level += ki[dst].level + 1;
                        if (maxlvl < ki[n].level)
                                maxlvl = ki[n].level;
                }

                while (nsrc) {
                        if (src < dst) {
                                kn = ki[src];
                                memmove(ki + src, ki + src + 1,
                                    (dst - src + ndst - 1) * sizeof *ki);
                                ki[dst + ndst - 1] = kn;
                                nsrc--;
                                dst--;
                                ndst++;
                        } else if (src != dst + ndst) {
                                kn = ki[src];
                                memmove(ki + dst + ndst + 1, ki + dst + ndst,
                                    (src - dst - ndst) * sizeof *ki);
                                ki[dst + ndst] = kn;
                                ndst++;
                                nsrc--;
                                src++;
                        } else {
                                ndst += nsrc;
                                src += nsrc;
                                nsrc = 0;
                        }
                }
        }
        /*
         * Now populate prefix (instead of level) with the command
         * prefix used to show descendancies.
         */
        path = calloc(1, (maxlvl + 7) / 8);
        if (path == NULL)
                err(1, NULL);

        for (src = 0; src < items; src++) {
                if ((lvl = ki[src].level) == 0) {
                        ki[src].prefix = NULL;
                        continue;
                }

                if ((ki[src].prefix = malloc(lvl * 2 + 1)) == NULL)
                        err(1, NULL);

                for (n = 0; n < lvl - 2; n++) {
                        ki[src].prefix[n * 2] =
                            path[n / 8] & 1 << (n % 8) ? '|' : ' ';
                        ki[src].prefix[n * 2 + 1] = ' ';

                }
                if (n == lvl - 2) {
                        /* Have I any more siblings? */
                        for (siblings = 0, dst = src + 1; dst < items; dst++) {
                                if (ki[dst].level > lvl)
                                        continue;
                                if (ki[dst].level == lvl)
                                        siblings = 1;
                                break;
                        }
                        if (siblings)
                                path[n / 8] |= 1 << (n % 8);
                        else
                                path[n / 8] &= ~(1 << (n % 8));
                        ki[src].prefix[n * 2] = siblings ? '|' : '`';
                        ki[src].prefix[n * 2 + 1] = '-';
                        n++;
                }
                strlcpy(ki[src].prefix + n * 2, "- ", (lvl - n) * 2 + 1);
        }
        free(path);
}

static void
usage(void)
{
        fprintf(stderr, "usage: %s [[-]AacefHhjkLlmrSTuvwxZ] [-M core]"
            " [-N system] [-O fmt] [-o fmt]\n", __progname);
        fprintf(stderr, "%-*s[-p pid] [-t tty] [-U user] [-W swap]\n",
            (int)strlen(__progname) + 8, "");
        exit(1);
}